### PR TITLE
feat: created some 3d-functionality.

### PR DESCRIPTION
**Pull request by MichaelLangbein**
Introduces a component `ukis-map-three`, analogous to `ukis-map-ol` but for a threejs backend, and also some raw WebGL custom-renderers.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/dlr-eoc/ukis-frontend-libraries/blob/master/CONTRIBUTING.md#-commit-message-guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] The Project is [building and passes all tests](https://github.com/dlr-eoc/ukis-frontend-libraries/blob/master/DEVELOPMENT.md#further-you-can-test-and-build-locally) without errors

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

Ukis has been designed with the intention of catering to different kinds of mapping-engines, not only focusing on openlayers but also allowing for projects like cesium, worldwind etc. This PR introduces threejs as a mapping library. 

The use-case for this particular library lies mostly in irregularly shaped objects. 
There are a few objects that are very big, but still irregularly shaped.
Because they are big, it makes sense to serve their image-material per WMTS instead of a single, humongous file.
Because they are irregularly shaped, it makes sense to display them in 3d.

## What is the new behavior?
A threejs scene is being created with orbit-control circling around a singular Mesh.
This mesh makes use of a custom shader, that uses a canvas of a map in EPSG:4326 as a texture. This way, any map in EPSG:4326 can be wrapped around any shape.
The texture-canvas is being kept in sync with the map. As such, when new tiles are loaded on the map, they are immediately visible on the mesh.

screenshot

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
Is it part of one/more [packages](https://github.com/dlr-eoc/ukis-frontend-libraries/packages) and which?
This PR  affects the projects demo-maps, map-three and utils-maps.

### CODE REVIEW

File: projects/demo-maps/src/app/route-components/route-example-custom-layers/route-map4.component.ts [hunk 1]
```diff
@@ -35,7 +33,9 @@ export class RouteMap4Component implements OnInit, AfterViewInit {
   constructor(
     public layersSvc: LayersService,
     public mapStateSvc: MapStateService,
-    public mapSvc: MapOlService) {
+    public mapSvc: MapOlService,
+    private zone: NgZone
```
Comment (boeckMt) [inline projects/demo-maps/src/app/route-components/route-example-custom-layers/route-map4.component.ts:37]: NgZone is not used please remove it.

File: projects/demo-maps/src/app/route-components/route-example-custom-layers/route-map4.component.ts [hunk 2]
```diff
@@ -1,30 +1,28 @@
-import { Component, OnInit, HostBinding, AfterViewInit, Inject } from '@angular/core';
+import { Component, OnInit, HostBinding, AfterViewInit, Inject, NgZone } from '@angular/core';
```
Comment (boeckMt) [inline projects/demo-maps/src/app/route-components/route-example-custom-layers/route-map4.component.ts:1]: Inject is not used please remove it.

File: projects/demo-maps/src/app/route-components/route-example-custom-layers/resources/features.ts [hunk 3]
```diff
@@ -0,0 +1,366 @@
+export const windMeasurements = {
```
Comment (boeckMt) [inline projects/demo-maps/src/app/route-components/route-example-custom-layers/resources/features.ts:1]: If you create a own file for this maybe we can outsource all data for Vector Layers?

File: projects/demo-maps/src/app/route-components/route-example-custom-layers/sunlight/sunlight.component.html [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<form clrForm [formGroup]="sunlightForm">
```
Comment (boeckMt) [inline projects/demo-maps/src/app/route-components/route-example-custom-layers/sunlight/sunlight.component.html:1]: If this is a normal UI component then it should be in the components folder.

File: projects/demo-maps/src/app/route-components/route-example-threejs/route-example-threejs.component.html [hunk 5]
```diff
@@ -0,0 +1,49 @@
+<main class="content-area">
```
Comment (boeckMt) [inline projects/demo-maps/src/app/route-components/route-example-threejs/route-example-threejs.component.html:1]: Please adjust the columns so it is better usable on smaller screens. (https://clarity.design/documentation/grid)
Two maps side by side is always a hard to layout but a lot of scrollbars look really ugly.

File: projects/demo-maps/src/app/route-components/route-example-threejs/route-example-threejs.component.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
+import { Component, OnInit, HostBinding, ViewChild, ElementRef, AfterViewInit, NgZone } from '@angular/core';
```
Comment (boeckMt) [inline projects/demo-maps/src/app/route-components/route-example-threejs/route-example-threejs.component.ts:1]: Also remove not used Imports.

File: projects/map-three/ng-package.json [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  },
+  "whitelistedNonPeerDependencies": [
+    "tslib",
+    "ol",
+    "proj4",
+    "@dlr-eoc/services-layers",
+    "@dlr-eoc/services-map-state",
+    "@dlr-eoc/services-map-state",
+    "@dlr-eoc/services-layers",
+    "@dlr-eoc/map-ol",
```
Comment (boeckMt) [inline projects/map-three/ng-package.json:15]: Does this Component needs map-ol?

File: projects/map-three/src/lib/map-three.component.scss [hunk 8]
```diff
@@ -0,0 +1,5 @@
+.threeCanvas {
+    width: 100%;
+    height: 600px;
```
Comment (boeckMt) [inline projects/map-three/src/lib/map-three.component.scss:3]: Always try to set sizes in procent not in pixel, otherwise the user has to overwrite it.

File: tsconfig.base.json [hunk 9]
```diff
@@ -33,6 +33,10 @@
       "utils-browser": [
```
Comment (boeckMt) [inline tsconfig.base.json:33]: All Paths here should be removed!
Only @dlr-eoc and ol should stay.

**Comment by boeckMt** [comment]
@MichaelLangbein this are a lot of changes in only one commit, it is better to commit stuff separately so maybe only parts can be integrated.
Also i would be happy if we dont integrate such big Files like Phobos_1_1000.glb, maybe it can be hosted somewhere else.

But all in all is this 3d functionality a usable feature for the future.

File: projects/demo-maps/src/app/route-components/route-example-custom-layers/resources/features.ts (hunk 3 above)
Comment (MichaelLangbein) [reply projects/demo-maps/src/app/route-components/route-example-custom-layers/resources/features.ts:1]: Sure thing! Will do!

**Comment by MichaelLangbein** [comment]
> this are a lot of changes in only one commit, it is better to commit stuff separately
Fair point! These are indeed quite a few changes. I have kept the original history in a separate branch - but I've dragged the branch along for about a month, working on it when there was a little time, and so with so many merges, complete overhauls and non-standard-conform commit-messages the history of it is now rather hard to follow: https://github.com/dlr-eoc/ukis-frontend-libraries/commits/custom_renderer.

I could instead merge that old branch, if you want, but considering the history gets squashed again anyway I feel there would be little use in that for us. 

Alternatively I could create some new commits, first removing all features and then adding them again one by one. Would that make more sense?

> Also i would be happy if we dont integrate such big Files like Phobos_1_1000.glb, maybe it can be hosted somewhere else.
Right! I'll search for another place to host this - can probably just fetch it from NASA's original page. Will update that in a bit.

**Comment by MichaelLangbein** [comment]
> > this are a lot of changes in only one commit, it is better to commit stuff separately
> 
> Fair point! These are indeed quite a few changes. I have kept the original history in a separate branch - but I've dragged the branch along for about a month, working on it when there was a little time, and so with so many merges, complete overhauls and non-standard-conform commit-messages the history of it is now rather hard to follow: https://github.com/dlr-eoc/ukis-frontend-libraries/commits/custom_renderer.
> 
> I could instead merge that old branch, if you want, but considering the history gets squashed again anyway I feel there would be little use in that for us.
>  make more sense?

Alright! History has been rebased, different features are now all in separate commits.

File: projects/demo-maps/src/app/route-components/route-example-custom-layers/route-map4.component.ts (hunk 2 above)
Comment (MichaelLangbein) [reply projects/demo-maps/src/app/route-components/route-example-custom-layers/route-map4.component.ts]: Done!

File: projects/demo-maps/src/app/route-components/route-example-custom-layers/sunlight/sunlight.component.html (hunk 4 above)
Comment (MichaelLangbein) [reply projects/demo-maps/src/app/route-components/route-example-custom-layers/sunlight/sunlight.component.html]: Done!

File: projects/demo-maps/src/app/route-components/route-example-custom-layers/resources/features.ts (hunk 3 above)
Comment (MichaelLangbein) [reply projects/demo-maps/src/app/route-components/route-example-custom-layers/resources/features.ts:1]: Done!

File: projects/demo-maps/src/app/route-components/route-example-threejs/route-example-threejs.component.html (hunk 5 above)
Comment (MichaelLangbein) [reply projects/demo-maps/src/app/route-components/route-example-threejs/route-example-threejs.component.html:1]: Done!

File: projects/demo-maps/src/app/route-components/route-example-threejs/route-example-threejs.component.ts (hunk 6 above)
Comment (MichaelLangbein) [reply projects/demo-maps/src/app/route-components/route-example-threejs/route-example-threejs.component.ts]: Done!

File: projects/map-three/ng-package.json (hunk 7 above)
Comment (MichaelLangbein) [reply projects/map-three/ng-package.json:14]: Done! Did need map-ol, but not ol.

File: projects/map-three/src/lib/map-three.component.scss (hunk 8 above)
Comment (MichaelLangbein) [reply projects/map-three/src/lib/map-three.component.scss]: Done!

File: tsconfig.base.json (hunk 9 above)
Comment (MichaelLangbein) [reply tsconfig.base.json]: Done!

**Review by boeckMt — APPROVED** [review]
